File: internal/handlers/wallet.go
package handlers

import (
	"database/sql"
	"encoding/json"
	"net/http"
	"time"

	"browser-server/internal/db"
	"browser-server/internal/helpers"
	"browser-server/internal/models"
)

func GetWallet(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")

	userID := helpers.GetUserIDFromQuery(r)
	websiteFilter := r.URL.Query().Get("website")

	query := "SELECT id, user_id, username, password, website, description, created_at, updated_at FROM wallet WHERE 1=1"
	args := []interface{}{}

	if userID > 0 {
		query += " AND user_id = ?"
		args = append(args, userID)
	}

	if websiteFilter != "" {
		query += " AND website LIKE ?"
		args = append(args, "%"+websiteFilter+"%")
	}

	rows, err := db.WalletDB.Query(query, args...)
	if err != nil {
		http.Error(w, "Database error", http.StatusInternalServerError)
		return
	}
	defer rows.Close()

	wallet := []models.WalletEntry{}
	for rows.Next() {
		var entry models.WalletEntry
		err := rows.Scan(&entry.ID, &entry.UserID, &entry.Username, &entry.Password, &entry.Website, &entry.Description, &entry.CreatedAt, &entry.UpdatedAt)
		if err != nil {
			continue
		}
		// Passwords stay hidden by default; they must be requested via the
		// reveal endpoint for a specific website + username.
		entry.Password = ""
		wallet = append(wallet, entry)
	}

	json.NewEncoder(w).Encode(wallet)
}

// RevealWalletPassword returns the password for a single credential identified
// by the requested user, website, and username. All three are required so a
// password can only be requested for a specific domain + username pair.
func RevealWalletPassword(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")

	userID := helpers.GetUserIDFromQuery(r)
	website := r.URL.Query().Get("website")
	username := r.URL.Query().Get("username")

	if userID == 0 || website == "" || username == "" {
		http.Error(w, "user_id, website, and username are required", http.StatusBadRequest)
		return
	}

	var password string
	err := db.WalletDB.QueryRow(
		"SELECT password FROM wallet WHERE user_id = ? AND website = ? AND username = ? LIMIT 1",
		userID, website, username,
	).Scan(&password)

	if err == sql.ErrNoRows {
		http.Error(w, "Wallet entry not found", http.StatusNotFound)
		return
	} else if err != nil {
		http.Error(w, "Database error", http.StatusInternalServerError)
		return
	}

	json.NewEncoder(w).Encode(map[string]string{"password": password})
}

func CreateWalletEntry(w http.ResponseWriter, r *http.Request) {
	var entry models.WalletEntry
	if err := json.NewDecoder(r.Body).Decode(&entry); err != nil {
		http.Error(w, "Invalid JSON", http.StatusBadRequest)
		return
	}

	result, err := db.WalletDB.Exec("INSERT INTO wallet (user_id, username, password, website, description) VALUES (?, ?, ?, ?, ?)",
		entry.UserID, entry.Username, entry.Password, entry.Website, entry.Description)
	if err != nil {
		http.Error(w, "Database error", http.StatusInternalServerError)
		return
	}

	id, _ := result.LastInsertId()
	entry.ID = int(id)
	entry.CreatedAt = time.Now()
	entry.UpdatedAt = time.Now()

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(entry)
}

func GetWalletByID(w http.ResponseWriter, r *http.Request) {
	id := helpers.GetIDFromPath(r)

	var entry models.WalletEntry
	err := db.WalletDB.QueryRow("SELECT id, user_id, username, password, website, description, created_at, updated_at FROM wallet WHERE id = ?", id).
		Scan(&entry.ID, &entry.UserID, &entry.Username, &entry.Password, &entry.Website, &entry.Description, &entry.CreatedAt, &entry.UpdatedAt)

	if err == sql.ErrNoRows {
		http.Error(w, "Wallet entry not found", http.StatusNotFound)
		return
	} else if err != nil {
		http.Error(w, "Database error", http.StatusInternalServerError)
		return
	}

	// Password stays hidden by default; use the reveal endpoint to request it.
	entry.Password = ""

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(entry)
}

func UpdateWalletEntry(w http.ResponseWriter, r *http.Request) {
	id := helpers.GetIDFromPath(r)

	var entry models.WalletEntry
	if err := json.NewDecoder(r.Body).Decode(&entry); err != nil {
		http.Error(w, "Invalid JSON", http.StatusBadRequest)
		return
	}

	_, err := db.WalletDB.Exec("UPDATE wallet SET user_id = ?, username = ?, password = ?, website = ?, description = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?",
		entry.UserID, entry.Username, entry.Password, entry.Website, entry.Description, id)

	if err != nil {
		http.Error(w, "Database error", http.StatusInternalServerError)
		return
	}

	entry.ID = id
	entry.UpdatedAt = time.Now()

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(entry)
}

func DeleteWalletEntry(w http.ResponseWriter, r *http.Request) {
	id := helpers.GetIDFromPath(r)

	result, err := db.WalletDB.Exec("DELETE FROM wallet WHERE id = ?", id)
	if err != nil {
		http.Error(w, "Database error", http.StatusInternalServerError)
		return
	}

	rowsAffected, _ := result.RowsAffected()
	if rowsAffected == 0 {
		http.Error(w, "Wallet entry not found", http.StatusNotFound)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

File: frontend/src/lib/api/wallet.ts
import type { WalletEntry, WalletImportResult } from '@browser-server/shared-types';
import { API_BASE, authHeaders } from './client';

export function getWallet(userId?: number, website?: string): Promise<WalletEntry[]> {
  const params = new URLSearchParams();
  if (userId) params.set('user_id', String(userId));
  if (website) params.set('website', website);
  const qs = params.toString();
  return fetch(`${API_BASE}/api/wallet${qs ? `?${qs}` : ''}`, { headers: authHeaders() }).then(
    (res) => {
      if (!res.ok) throw new Error(`Request failed: ${res.status}`);
      return res.json() as Promise<WalletEntry[]>;
    },
  );
}

export function revealWalletPassword(userId: number, id: number): Promise<string> {
  const params = new URLSearchParams({ user_id: String(userId), id: String(id) });
  return fetch(`${API_BASE}/api/wallet/reveal?${params.toString()}`, {
    headers: authHeaders(),
  }).then(async (res) => {
    if (!res.ok) {
      const text = await res.text();
      throw new Error(text || `Request failed: ${res.status}`);
    }
    return ((await res.json()) as { password: string }).password;
  });
}

export function importWallet(userId: number, file: File): Promise<WalletImportResult> {
  const formData = new FormData();
  formData.append('file', file);
  return fetch(`${API_BASE}/api/wallet/import?user_id=${userId}`, {
    method: 'POST',
    headers: authHeaders(),
    body: formData,
  }).then(async (res) => {
    if (!res.ok) {
      const text = await res.text();
      throw new Error(text || `Import failed: ${res.status}`);
    }
    return res.json() as Promise<WalletImportResult>;
  });
}

export function getWalletEntry(id: number): Promise<WalletEntry> {
  return fetch(`${API_BASE}/api/wallet/${id}`, { headers: authHeaders() }).then((res) => {
    if (!res.ok) throw new Error(`Request failed: ${res.status}`);
    return res.json() as Promise<WalletEntry>;
  });
}

export function createWalletEntry(data: {
  user_id: number;
  website: string;
  username: string;
  password: string;
  login_provider?: string;
  description?: string;
}): Promise<WalletEntry> {
  return fetch(`${API_BASE}/api/wallet`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...authHeaders() },
    body: JSON.stringify(data),
  }).then((res) => {
    if (!res.ok) throw new Error(`Request failed: ${res.status}`);
    return res.json() as Promise<WalletEntry>;
  });
}

export function updateWalletEntry(id: number, data: Partial<WalletEntry>): Promise<WalletEntry> {
  return fetch(`${API_BASE}/api/wallet/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json', ...authHeaders() },
    body: JSON.stringify(data),
  }).then((res) => {
    if (!res.ok) throw new Error(`Request failed: ${res.status}`);
    return res.json() as Promise<WalletEntry>;
  });
}

export function deleteWalletEntry(id: number): Promise<void> {
  return fetch(`${API_BASE}/api/wallet/${id}`, { method: 'DELETE', headers: authHeaders() }).then(
    (res) => {
      if (res.status === 204) return;
      if (!res.ok) throw new Error(`Request failed: ${res.status}`);
    },
  );
}
